protocol/op-stack-manager-arch.md

```diff
@@ -0,0 +1,292 @@
+- Milestone 3: Supporting interop upgrades.
+
+This design doc proposes the architecture for a Milestone 0.5 of "OPSM is used for all interop devnet & testnet L1 deployments".
+Adding this milestone will allow the interop team to build testing infrastructure that wraps OPSM, unifying our development, testnet, and production testing and deployment processes.
+
+# Problem Statement + Context
+
+*Some text copied from https://github.com/ethereum-optimism/design-docs/pull/52, which is related to this design doc.*
+
+The current L2 chain deployment approach originates from a time with Hardhat, single L1 target, and a single monolithic set of features.
+Since then the system has migrated to Foundry and extended for more features, but remains centered around a single monolithic deploy-config for all its features.
+
+The interop team needs a way to configure new multi-L2 deployments: The number of ways to compose L2s in tests grows past what a single legacy config template can support.
+
+Outside of interop, deployment also seems increasingly complex and opaque, while it does not have to be, due to the same configuration and composability troubles.
+
+Part of the solution to this is described in https://github.com/ethereum-optimism/design-docs/pull/52,and the other part involves restructuring our Solidity config processing and deploy scripts.
+This design doc focuses on the L1 contract configuration and deployment part of the solution.
+
+# Alternatives Considered
+
+The primary alternative is for the interop team to not use OPSM for devnet and testnet deployments.
+This is not ideal for various reasons:
+
+- Duplication of effort: The interop team would need to develop and maintain their own deployment scripts, even though new ones will already be written for OPSM.
+- Divergence: With duplication, it's likely the interop team's deployment scripts diverge from the production deployment scripts, leading to inconsistencies and bugs.
+
+# Proposed Solution
+
+There are two steps to deploying L1 contracts:
+
+1. **Deploy Superchain Contracts**. This only occurs occasionally, as the Superchain contracts are shared between many OP chains.
+2. **Deploy Shared Implementation Contracts**. This occurs once per [contracts release](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/VERSIONING.md).
+3. **Deploy OP Chain Contracts**. This occurs for once every OP chain deployment.
+
+For each step we define the step inputs (configuration), sub-steps that occur, and the step outputs (artifacts).
+For each we start by listing the sub-steps, then define the inputs and outputs for each sub-step, and use those to define the inputs and outputs for the step as a whole.
+
+Sample input and output TOML files are included for each step.
+The exact structure of these files is too low-level to be in-scope for this design doc, so just consider them as examples.
+However, feel free to comment on specifics if you think they're important.
```

Comment:
I am fine with TOML, but I'd love some config abstraction, and not directly bundle TOML logic into deploy scripts. We can do small config contracts for usage in scripts, with a default implementation of the config interface that uses the TOML files. Having this interface can make the integration into Go tooling / testing a lot smoother, by not having to go through intermediate scripts on disk.

Reply:
Got it. I'm assuming this is expanded on more in https://github.com/ethereum-optimism/design-docs/pull/61, so I'll read that next

Reply:
decision: Enable passing config in-process (e.g. `Config` struct) and via input files